server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSource.java

```diff
@@ -58,4 +66,69 @@ void getSegmentFiles(
      * Cancel any ongoing requests, should resolve any ongoing listeners with onFailure with a {@link ExecutionCancelledException}.
      */
     default void cancel() {}
+
+    /**
+     * Directory wrapper that records copy process for replication statistics
+     *
+     * @opensearch.internal
+     */
+    final class ReplicationStatsDirectoryWrapper extends FilterDirectory {
```

Comment:
@andrross Thanks for this suggestion. Tested the changes - we have proper progress tracking now, instead of just 0% or 100%.